fix(rust,python): clean up `dt` namespace, fix `mean`/`median` return dtypes, impl `std` for `date`/`datetime`

mcrumiller opened the pull request:
**This is a breaking change**

Resolves #10232, #12870, and #12883. @MarcoGorelli two notes to consider:

1. The breaking change here is that the `mean` and `median` functions for `pl.Date` now return a `datetime.datetime` instead of a `datetime.date`. This seemed much more logical if we consider dates corresponding to integers and datetimes corresponding to floats, since the mean of `[1, 2]` is 1.5, so too is the mean of `[date(2023, 1, 1), date(2023, 1, 2)]` now `datetime(2023, 1, 1, 12, 0, 0)`. Let me know if this seems reasonable. If not, it's easy to switch to return dates.
2. The Series `dt` namespace has a defined `mean/median` but the Expr `dt` namespace does not. I don't see why we should have special mean and median functions defined in the `dt` namespace at all, as we should be able to simply call `series.mean()` and it should return the appropriate result. I set this up so that calling `s.mean()` simply dispatches to the `s.dt.mean()`, but I think we should perhaps deprecate `.dt.mean/median`, what do you think?

I also implemented this for the duration types.

stinodego commented on crates/polars-core/src/series/implementations/dates_time.rs:
```diff
@@ -172,7 +172,8 @@ macro_rules! impl_dyn_series {
 
             fn mean(&self) -> Option<f64> {
-                self.0.mean()
+                let s = self.0.mean();
+                s
```
This syntax is less readable than the original, I'd recommend reverting this.

mcrumiller replied:
Ahh yes, I did this during debugging and did not intend to keep this.

stinodego reviewed (changes requested):
I'll defer to @MarcoGorelli on this one.

My initial thoughts are: 
* Returning a `Datetime` rather than a `Date` makes sense - I can see the parallel with ints/floats. Though it really is a different data type so we have to make sure this works for large date values.
* I don't really see the benefit of dedicated namespace methods. We can probably deprecate `dt.mean` in favor of `mean`.

MarcoGorelli commented on crates/polars-core/src/series/implementations/dates_time.rs:
```diff
@@ -359,10 +360,10 @@ macro_rules! impl_dyn_series {
-                    .into()
+                Series::new(
+                    self.name(),
+                    &[self.median().map(|v| (v * 86_400_000_000f64) as i64)],
```
can you use one of the constants here? like `US_IN_DAY`?

MarcoGorelli commented on crates/polars-core/src/series/mod.rs:
```diff
@@ -805,7 +805,13 @@ impl Series {
-            dt @ DataType::Duration(_) => {
+            DataType::Date => Series::new(
+                self.name(),
+                &[self.mean().map(|v| (v * 86_400_000_000f64) as i64)],
```
same

MarcoGorelli commented on py-polars/polars/series/datetime.py:
```diff
@@ -80,15 +83,18 @@ def median(self) -> dt.date | dt.datetime | dt.timedelta | None:
             if s.dtype == Date:
-                return _to_python_date(int(out))
-            else:
+                # 86_400_000_000 = microseconds in days
+                return _to_python_datetime(int(out * 86_400_000_000), "us")
```
same, is there a constant on the python side?

mcrumiller replied:
We have `NS_PER_SECOND` AND `SECONDS_PER_DAY` so I brought those in to create a `US_PER_DAY`. Thanks for the suggestions.

MarcoGorelli reviewed (commented):
generally on board with the idea

note that for `interpolate`, `Date` is also preserved:
```python
In [23]: pl.Series([date(2020, 1, 1), None, date(2020, 1, 2)]).interpolate()
Out[23]:
shape: (3,)
Series: '' [date]
[
        2020-01-01
        2020-01-01
        2020-01-02
]
```
, maybe that should also become `Datetime[us, None]`? (that's for a separate PR though)

Also agree with deprecating `.dt.mean` (but, also, that's for a separate PR)

mcrumiller commented:
> I'll defer to @MarcoGorelli on this one.
> 
> My initial thoughts are:
> 
> * Returning a `Datetime` rather than a `Date` makes sense - I can see the parallel with ints/floats. Though it really is a different data type so we have to make sure this works for large date values.
> * I don't really see the benefit of dedicated namespace methods. We can probably deprecate `dt.mean` in favor of `mean`.

@stinodego should I move the date-specific logic into `series.py` then, or should we keep the logic in `datetime.py` and import it, but keep it outside of the `.dt` namespace? In other words, is `datetime.py` *only* reserved for the `dt` namespace, or should it also hold date-specific logic?

mcrumiller commented:
> * Returning a `Datetime` rather than a `Date` makes sense - I can see the parallel with ints/floats. Though it really is a different data type so we have to make sure this works for large date values.

Both the `mean` and `median` functions map from `i32` or `i64` (date vs datetime) to `f64` to perform the mean anyway. Since `datetime` is backed by `i64` whereas dates are only backed by `i32`, I think there's no issue with date -> datetime conversion: even with the finest time unit of nanoseconds, our maximum datetime is $\frac{2^{63}-1}{1,000,000,000}=9,223,372,036$ days since the unix epoch, compared to `dates` which have a max date of $2^{31}-1=4,294,967,295$ days.

MarcoGorelli reviewed (changes requested):
thanks for updating - not sure about the if/then statements in `Series.mean` and co, will take a closer look

mcrumiller commented on py-polars/polars/series/series.py:
```diff
@@ -1672,7 +1672,10 @@ def mean(self) -> int | float | None:
 
         """
-        return self._s.mean()
+        if self.dtype.is_temporal():
+            return self.dt.mean()
+        else:
+            return self._s.mean()
```
Hye @MarcoGorelli give me a bit, I'm pushing this into the rust side so we can have a cleaner call to `s.mean()` from the python side.

mcrumiller commented:
@MarcoGorelli I just pushed a new version with the following:
 
* Updated the `mean` and `median` on the rust side so for temporal dtypes it returns the correct dtype directly back to python
* Removed `.dt.mean`, `median`, `min`, and `max` from the `dt` namespace, as these function properly via the normal series functions.
* Removed if/then temporal code from `series.py`.

mcrumiller commented:
Ahh I should deprecate instead of move.

mcrumiller commented:
Not sure if the test failure is related.

mcrumiller commented:
Thanks @stinodego. I went ahead and added `std` for `date`/`datetime` series for issue #12883 as well since it seems to fit in here.

mcrumiller commented on py-polars/polars/series/datetime.py:
```diff
@@ -29,87 +33,25 @@ def __getitem__(self, item: int) -> dt.date | dt.datetime | dt.timedelta:
+
+    @deprecate_function("Use `series.median` instead.", version="0.19.20")
+    def median(self) -> dt.datetime | dt.timedelta | None:
+        """Return median as python DateTime."""
```
Sounds resonable. `pl.Date` is no longer returned but that sounds nitpicky. I also need to add `pl.Time`. Have new PR incoming.

MarcoGorelli commented:
> Both the `mean` and `median` functions map from `i32` or `i64` (date vs datetime) to `f64` to perform the mean anyway. Since `datetime` is backed by `i64` whereas dates are only backed by `i32`, I think there's no issue with date -> datetime conversion: even with the finest time unit of nanoseconds, our maximum datetime is 263−11,000,000,000=9,223,372,036 days since the unix epoch, compared to `dates` which have a max date of 231−1=4,294,967,295 days.

this calculation is wrong :) for nanoseconds, you need to divide by `1_000_000 * 3600`, which does indeed give a smaller range than `Date` dtype

BUT

in Chrono, both 'us' and 'Date' are limited to dates between +/- 262,000 years since the unix epoch, so in practice, it's not a concern
